src/firebolt_db/firebolt_async_dialect.py

```diff
@@ -0,0 +1,141 @@
+        self.await_ = adapt_connection.await_
+        self.arraysize = 1
+        self.rowcount = -1
+        self.description = None
+        self._rows: List[Tuple] = []
+
+    def close(self) -> None:
+        self._rows[:] = []
+
+    def execute(self, operation: str, parameters: Optional[Tuple] = None) -> None:
+        _cursor = self._connection.cursor()
+        self.await_(_cursor.execute(operation, parameters))
+        if _cursor.description:
+            self.description = _cursor.description
+            self.rowcount = -1
+            self._rows = self.await_(_cursor.fetchall())
+        else:
+            self.description = None
+            self.rowcount = _cursor.rowcount
+
+        _cursor.close()
+
+    def executemany(self, operation: str, seq_of_parameters: List[Tuple]) -> None:
+        raise NotImplementedError("executemany is not supported yet")
+
+    def __iter__(self) -> Iterator[Tuple]:
+        while self._rows:
+            yield self._rows.pop(0)
+
+    def fetchone(self) -> Optional[Tuple]:
+        if self._rows:
+            return self._rows.pop(0)
+        else:
+            return None
+
+    def fetchmany(self, size: int = None) -> List[Tuple]:
+        if size is None:
+            size = self.arraysize
+
+        retval = self._rows[0:size]
+        self._rows[:] = self._rows[size:]
+        return retval
+
+    def fetchall(self) -> List[Tuple]:
+        retval = self._rows[:]
+        self._rows[:] = []
+        return retval
+
+
+class AsyncConnectionWrapper(AdaptedConnection):
+    await_ = staticmethod(await_only)
+    __slots__ = ("dbapi", "_connection")
+
+    def __init__(self, dbapi: AsyncAPIWrapper, connection: Connection):
+        self.dbapi = dbapi
+        self._connection = connection
+
+    def cursor(self) -> AsyncCursorWrapper:
+        return AsyncCursorWrapper(self)
+
+    def rollback(self) -> None:
+        pass
+
+    def commit(self) -> None:
+        self._connection.commit()
+
+    def close(self) -> None:
+        self.await_(self._connection._aclose())
+
+
+class AsyncAPIWrapper:
+    """Wrapper around Firebolt async dbapi that returns a similar wrapper for
+    Cursor on connect()"""
+
+    def __init__(self, dbapi: ModuleType):
+        self.dbapi = dbapi
+        self.paramstyle = dbapi.paramstyle  # type: ignore[attr-defined] # noqa: F821
+        self._init_dbapi_attributes()
+
+    def _init_dbapi_attributes(self) -> None:
+        for name in (
+            "DatabaseError",
+            "Error",
+            "IntegrityError",
+            "NotSupportedError",
+            "OperationalError",
+            "ProgrammingError",
+        ):
+            setattr(self, name, getattr(self.dbapi, name))
+
+    def connect(self, *arg: Any, **kw: Any) -> AsyncConnectionWrapper:
+
+        connection = self.dbapi.connect(*arg, **kw)  # type: ignore[attr-defined] # noqa: F821,E501
+        return AsyncConnectionWrapper(
+            self,
+            await_only(connection),
```

Comment:
why do we await here and not when we do `self.dbapi.connect`?

Reply:
No particular reason apart from seeing something similar in another example. If you feel like doing it on L123 would be clearer I can move it around.

Reply:
I was just curious. It just was confusing that `connection` variable wasn't indeed a connection, but a coroutine

Reply:
fair point, refactored it to be clearer.